### PR TITLE
Fix conflict between Sys and SysBasic definitions of off_t, mode_t, socklen_t

### LINKED ISSUE #18173: Support renaming extern types
Most Chapel `extern` declarations support a means of giving the C name of the symbol a separate identifier from the Chapel name, which is particularly useful with the C name isn't a legal Chapel identifier.  For instance:

```chapel
extern "domain" var c_domain: c_int;
```

However, it appears that `extern type` declarations don't support this today.  I think they should for completeness and orthogonality, and that this is just an oversight / a case that nobody has tackled yet.  I also believe this would be a relatively straightforward project for someone new to the code base and interested in modifying the compiler (and would be happy to advise / assist).

### LINKED ISSUE #19279: Using types moved from `SysBasic` to `Sys` breaks when both modules are used
In https://github.com/chapel-lang/chapel/pull/19262#issuecomment-1045082400, @mppf pointed out that my move of types like `socklen_t` from `SysBasic` to `Sys` breaks programs that use both since I relied on a `proc ... type` to implement the warning that the type had moved.  Specifically, a program like:

```chapel
use Sys;
use SysBasic;

var x: socklen_t;
```

results in:

```
$CHPL_HOME/modules/standard/Sys.chpl:258: error: symbol socklen_t is multiply defined
$CHPL_HOME/modules/standard/SysBasic.chpl:78: note: also defined as a function here (and possibly elsewhere)
```

This is arguably less of an issue for types that moved from `SysBasic` to `CTypes` since existing programs would not use both modules since `CTypes` had not existed before this point, and would hopefully move over to `CTypes` and stop using `SysBasic` (unless they relied on other elements of `SysBasic` in which case they'd need to use an `only` clause to get at them in order to not hit this same issue).

**Associated Future Test(s):**
[`test/library/packages/canCompileNoLink/testSockets.chpl`](
           https://github.com/chapel-lang/chapel/blob/main/test/library/packages/canCompileNoLink/testSockets.chpl
          ) #19345

### PR DESCRIPTION
**Pull request by bradcray**
This PR deprecates the types:

* off_t
* mode_t
* socklen_t

replacing them with c_off_t, c_mode_t, and c_socklen_t, as proposed
in https://github.com/chapel-lang/chapel/issues/18012.  This has the
beneficial side-effect of resolving an issue that I introduced in
https://github.com/chapel-lang/chapel/pull/19262 in which my trick for
forwarding these types from SysBasic to Sys caused problems when both
modules were 'use'd because it had the effect of bringing a variable and
procedure overload into the same scope which the compiler doesn't
permit.

Replaced the forwarding procedures with deprecated types now that we're
using different types and updated the deprecation test's output:
* modules/standard/SysBasic.chpl
* test/deprecated/sysBasicTypes.good

Simplified the helper routine in CTypes that I introduced to help with
these symbols renamings since it no longer needs to handle multiple
modules.
* modules/standard/CTypes.chpl

Updated tests to reflect new module inter-dependencies:
* test/compflags/ferguson/print-module-resolution.good
* test/modules/sungeun/init/printModuleInitOrder*.good

Updated uses of these types in tests and modules.
* modules/packages/HDF5.chpl
* modules/packages/Socket.chpl
* modules/standard/FileSystem.chpl
* modules/standard/Sys.chpl
* test/extern/kbrady/rename_extern_record.chpl

Added a new test that captures the case that had not worked
properly before.
* test/deprecated/SysBasicAndSys.chpl
* test/deprecated/SysBasicAndSys.good

Also add another tests that compiles the Sockets
package, as it would've caught this earlier, and will also do some
minimal testing of the Sockets package as we seem to have little
to none today.
* test/library/packages/canCompileNoLink/testSockets.chpl
* test/library/packages/canCompileNoLink/testSockets.good

As a semi-related side project, I also added support for renaming of
extern types to the old C parser because I thought it would be a
useful way to introduce these new type aliases without threading them
through the runtime.  However, I have not yet taught the new parser
how to read renamed extern types, so for now I am threading the
renamings through the runtime.
* compiler/AST/build.cpp
* compiler/include/build.h
* compiler/parser/chapel.ypp
* runtime/include/chpltypes.h

Added a new test of this extern type renaming:
* test/extern/bradc/types/cint-rename.chpl
* test/extern/bradc/types/cint-rename.good

TODO:
- [ ] re-generate parser files post-review

Resolves #19279.
Resolves #18173

### CODE REVIEW

**Comment by bradcray** [comment]
@mppf: Would you mind reviewing this, having pointed out the flaw in #19279 and reviewed #19262?

**Comment by bradcray** [comment]
Hmm, got a few failures to fix up still, so no rush.  Though they shouldn't take long either.

**Comment by bradcray** [comment]
@mppf: As an update, when putting the new `c_` types into the modules with the renaming, I forgot that I'd need to teach the back-end about them, so thought I'd bite the bullet and rename the long-overdue `extern "c_name" type` feature #18173 which was as easy as expected.  Except that I forgot I'd need to update the new front-end to support this as well, and could probably benefit from some pointers about how to do that.  Alternatively, I could introduce these types to chpltypes.h which is the simpler but less satisfying approach...

**Comment by bradcray** [comment]
@mppf: I think this could be merged as-is and we could update the dyno parser for this case later; or we could update the dyno parser to handle renaming of extern types as part of this PR as well.  I got as far as seeing that other extern renaming cases store the C name into a parser-local variable, but couldn't figure out what happened to it when building up the AST in the time I spent on it in order to apply a similar logic to the `TEXTERN STRINGLITERAL TTYPE` case.  As a result, I did end up adding the types to chpltypes.h for the time being for this PR.

**Review by mppf — APPROVED** [review]
Sounds good to me. In terms of updating the dyno parser, I think we can do that in a separate effort (and I don't think you need to do it).

In terms of `c_off_t`, `c_mode_t`, and `c_socket_t` - AFAIK, the versions of these in the `OS.POSIX` module (see #18448) will not have the `c_` to start with. It is not obvious to me that all of these belong in `OS.POSIX` with the C prefix - if they have the C prefix, maybe they belong in `CTypes`? And if they are in `OS.POSIX` then they would be `off_t`, `mode_t`, and `socket_t`? Maybe it depends on whether we consider them being part of the C standard or POSIX? Some Googling indicates that these are POSIX and not C standard things. Nonetheless I think that what we have here in this PR is OK for now and we might shuffle them more in the future. (If we want to limit renamings for users then I think we should try to get the `OS.POSIX` changes going this release; otherwise what we have here is reasonable as an interim solution IMO).

Lastly, as you know, we are discussing whether or not the standard library should be considered to be at a further scope in #19167. That would have avoided the original issue. Also, in my draft implementation work related to that issue, I enabled `pragma "last resort"` to be used for things resolved during scope resolve (variables/parenless functions) which would also help in this specific case.

File: compiler/include/build.h [hunk 1]
```diff
@@ -196,7 +196,7 @@ BlockStmt* buildSyncStmt(Expr* stmt);
 BlockStmt* buildCobeginStmt(CallExpr* byref_vars, BlockStmt* block);
 BlockStmt* buildExternBlockStmt(const char* c_code);
 CallExpr*  buildPreDecIncWarning(Expr* expr, char sign);
-BlockStmt* convertTypesToExtern(BlockStmt*);
+BlockStmt* convertTypesToExtern(BlockStmt*, const char* cname=NULL);
```
[the remark this answers is not in the document]
Comment (bradcray) [reply compiler/include/build.h:199]: Is there scripting to help catch such things, or are we just relying on code review to catch it?

**Comment by bradcray** [comment]
@mppf: Just getting back to this.

>  It is not obvious to me that all of these belong in OS.POSIX with the C prefix

I'd agree with that.

> - if they have the C prefix, maybe they belong in CTypes?

I don't think these types are primary/major/central enough to warrant being in CTypes, personally.

Your comments make me think maybe we don't like option 4 in https://github.com/chapel-lang/chapel/issues/19279 as much as we thought.  Here are some other options:

* go back to putting the types back in SysBasic.chpl for now (option 1): main downside is that it delays the inevitable, but it also reduces churn

* simply deprecate the types without offering an alternative.  This would require the small number of modules and tests that rely on these types to provide their own extern types for the time being / until Sys.POSIX exists.  There aren't many in our code base:

* modules/standard/FileSystem.chpl
* modules/packages/HDF5.chpl
* modules/packages/Socket.chpl
* modules/standard/Sys.chpl
* test/extern/kbrady/rename_extern_record.chpl

and I'd exist not many in user code either.  Arkouda does more system stuff (I/O, ZMQ) than most user apps and doesn't use any of them, e.g.

File: compiler/include/build.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (mppf) [reply compiler/include/build.h:199]: Today, just code review. The earlier effort in this area was PR #17167 which was using a `-W` flag to warn but there is some complexity around that due to the generated flex/bison code.

**Comment by mppf** [comment]
> Your comments make me think maybe we don't like option 4 in https://github.com/chapel-lang/chapel/issues/19279 as much as we thought.

I think that's a reasonable assessment. Earlier I was thinking Option 4 was doing a next step we were going to do anyway, but now I am less sure of that.

PR #19306 happens to implement Option 3 (scope resolve considers `pragma "last resort"`) and that change itself didn't seem particularly hard or scary (at least, to get it working well enough for these deprecation cases). I can pull that change out into a separate PR if you are interested in proceeding in that direction.

Another option that would be OK with me would be to keep these types in Sys but remove the deprecation warnings from SysBasic. I think doing so is justifiable on the argument that they aren't used very much in practice and so the deprecation warning might not be worth the trouble.

File: compiler/include/build.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (bradcray) [reply compiler/include/build.h:199]: OK, I have no memory of that discussion and generally rankle at being held to things that we don't have testing for, but will make the change.

**Comment by bradcray** [comment]
> PR https://github.com/chapel-lang/chapel/pull/19306 happens to implement Option 3 (scope resolve considers pragma "last resort") and that change itself didn't seem particularly hard or scary (at least, to get it working well enough for these deprecation cases). I can pull that change out into a separate PR if you are interested in proceeding in that direction.

I think that approach could be attractive, though I don't know that it needs to be pulled out into its own PR (unless we don't proceed with #19306 before the release, I suppose).

**Comment by bradcray** [comment]
Closing this, as @mppf got "last resort" working in #19372, which obviates the need for the approaches taken here, and I broke the other changes rolled into this PR into their own PRs, now merged:

* #19337
* #19345

**Pull request closed without merging**